snuba/query/processors/hash_map_optimizer.py

```diff
@@ -0,0 +1,192 @@
+            match is not None
+            and match.string(KEY_COL_MAPPING_PARAM) == f"{self.__column_name}.key"
+        ):
+            rhs = match.expression("right_hand_side")
+            assert isinstance(rhs, LiteralExpr)
+            return (
+                ConditionClass.NOT_OPTIMIZABLE
+                # ifNull(tags[asd], '') = '' is not optimizable.
+                if rhs.value == ""
+                else ConditionClass.OPTIMIZABLE
+            )
+        elif match is None:
+            # If this condition is not matching an optimizable condition,
+            # check that it does not reference the optimizable column.
+            # If it does, it means we should not optimize this query.
+            for exp in condition:
+                if isinstance(exp, Column) and exp.column_name in (
+                    f"{self.__column_name}.key",
+                    f"{self.__column_name}.value",
+                ):
+                    return ConditionClass.NOT_OPTIMIZABLE
+            return ConditionClass.IRRELEVANT
+        else:
+            return ConditionClass.IRRELEVANT
+
+    def __replace_with_hash_map(self, condition: Expression) -> Expression:
+        match = self.__optimizable_pattern.match(condition)
+        if (
+            match is None
+            or match.string(KEY_COL_MAPPING_PARAM) != f"{self.__column_name}.key"
+        ):
+            return condition
+        rhs = match.expression("right_hand_side")
+        assert isinstance(rhs, LiteralExpr)
+        return FunctionExpr(
+            alias=condition.alias,
+            function_name="has",
+            parameters=(
+                Column(
+                    alias=None,
+                    table_name=match.optional_string(TABLE_MAPPING_PARAM),
+                    column_name=self.__hash_map_name,
+                ),
+                FunctionExpr(
+                    alias=None,
+                    function_name="cityHash64",
+                    parameters=(
+                        LiteralExpr(
+                            None, f"{match.string(KEY_MAPPING_PARAM)}={rhs.value}"
+                        ),
+                    ),
+                ),
+            ),
+        )
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        if not get_config("tags_hash_map_enabled", 0):
+            return
+
+        cond_class = ConditionClass.IRRELEVANT
+        condition = query.get_condition_from_ast()
+        if condition is not None:
+            cond_class = self.__classify_combined_conditions(condition)
+            if cond_class == ConditionClass.NOT_OPTIMIZABLE:
+                return
+
+        having_cond_class = ConditionClass.IRRELEVANT
+        having_cond = query.get_having_from_ast()
+        if having_cond is not None:
+            having_cond_class = self.__classify_combined_conditions(having_cond)
+            if having_cond_class == ConditionClass.NOT_OPTIMIZABLE:
+                return
```

Comment:
Even if HAVING contained a non optimizable condition, would it still be potentially worthwhile to optimize the WHERE clause?

Reply:
This is a good question. The goal with NOT_OPTIMIZABLE is not to load the hashmap column if I am already loading the full tags column for all records. Most likely if there is a tags expression in the having clause, it is either done to do aggregations or to do the group by, which already means we are potentially loading it for all the rows for the aggregations. It may not be always the case but it certainly can happen. We do not have a lot of information at this point of query processing to make a more informed decision so I went for the safe option.
I think we could test changing this switch later on.